src/main/java/com/pusher/client/channel/impl/PrivateEncryptedChannelImpl.java

```diff
@@ -0,0 +1,146 @@
+package com.pusher.client.channel.impl;
+
+import com.google.gson.Gson;
+import com.pusher.client.AuthorizationFailureException;
+import com.pusher.client.Authorizer;
+import com.pusher.client.channel.PrivateEncryptedChannel;
+import com.pusher.client.channel.PrivateEncryptedChannelEventListener;
+import com.pusher.client.channel.SubscriptionEventListener;
+import com.pusher.client.connection.impl.InternalConnection;
+import com.pusher.client.crypto.nacl.SecretBoxOpener;
+import com.pusher.client.util.Base64;
+import com.pusher.client.util.Factory;
+
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public class PrivateEncryptedChannelImpl extends ChannelImpl implements PrivateEncryptedChannel {
+
+    private static final Gson GSON = new Gson();
+    private final InternalConnection connection;
+    private final Authorizer authorizer;
+    private SecretBoxOpener secretBoxOpener;
+
+    private class PrivateEncryptedChannelData {
+        byte[] auth;
+        final String channelData;
+
+        protected PrivateEncryptedChannelData(String auth, String channelData) {
+            this.auth = auth.getBytes();
+            this.channelData = channelData;
+        }
+
+        public String getAuth() {
+            return new String(auth);
```

Comment:
is this part okay `new String(auth);`?

Reply:
I'll let Marek speak to that.

Personally, I would keep it as a byte array everywhere, converting to a string only at the last moment that's necessary (during serialisation), and only if the GSON serialiser can't handle a byte array correctly...

Reply:
Hmm. So the GSON serialiser struggled when I initially tried - so I think i've improved that the `PrivateEncryptedChannelData` object expects a byte array and the other sides need to handle that.